### PR TITLE
EZP-29096: Do not store empty draft values in DB

### PR DESCRIPTION
**Pull request by pkamps**
The issue and steps to reproduce it are described here:
https://jira.ez.no/browse/EZP-29096

Or use those steps:

* Create a new user account
* Do not fill any values in the form
* Close the browser tab
* Try to create a new user account
* You will get a transaction error

This patch will prevent the user creation to store any draft data into the DB unless the editor specified a user login.

### CODE REVIEW

**Comment by andrerom** [comment]
Nice.

We should aim to place this one 2017.12 and up. As for Travis, unsure why but tests seems to fail with:
```
.PHP Fatal error:  Call to a member function store() on null in /home/travis/build/ezsystems/ezpublish-legacy/kernel/classes/datatypes/ezuser/ezusertype.php on line 282
```

File: kernel/classes/datatypes/ezuser/ezusertype.php [hunk 1]
```diff
@@ -258,7 +258,10 @@ function storeObjectAttribute( $contentObjectAttribute )
             }
 
             // saving information in the object attribute data_text field to simulate a draft
-            $contentObjectAttribute->setAttribute( 'data_text', $this->serializeDraft( $user ) );
+            if( $user->Login )
```
Comment (glye) [inline kernel/classes/datatypes/ezuser/ezusertype.php]:
```suggestion
            if ( $user->Login )
```

**Comment by glye** [comment]
onPublish() fails because there is no user object. Seems the code expects the draft to have been stored, so bigger changes are needed.
https://github.com/ezsystems/ezpublish-legacy/pull/1417/files#diff-7c45a8ca23a28df173a64ef5548180c0R282

**Comment by peterkeung** [comment]
Are tests passing with the latest commit?

**Comment by glye** [comment]
@peterkeung No, sorry. Don't you have access to see the results? Travis says `eZUserTypeRegression::testUpdatePasswordUpdatesSerializedData
The password hash stored in the user attribute should have been updated
Failed asserting that null is not equal to null.`

File: kernel/classes/datatypes/ezuser/ezuser.php [hunk 2]
```diff
@@ -225,15 +225,24 @@ static function create( $contentObjectID )
         return new eZUser( $row );
     }
 
-    function store( $fieldFilters = null )
+    /**
+     * Only stores the entry if it has a Login value
+     *
+     * @param null $fieldFilters
```
Comment (glye) [inline kernel/classes/datatypes/ezuser/ezuser.php]:
```suggestion
     * @param mixed|null $fieldFilters
```
Not much point in a param that is always null :)

File: tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php [hunk 3]
```diff
@@ -110,8 +115,9 @@ public function testFromStringHandlesDisabledAccount()
         $userAccount = $dataMap['user_account']->fromString(
             join( '|', array( $tmpLogin, $tmpEmail, self::USER_PASSWORD_HASH, self::USER_PASSWORD_HASH_ID, 0 ) )
         );
-        $userAccount->store();
 
+        $dataMap['user_account']->setContent( $userAccount );
+        $dataMap['user_account']->store();
```
Comment (glye) [inline tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php]: Is this change needed, or cosmetic? :)

File: tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php [hunk 4]
```diff
@@ -188,11 +195,13 @@ public function testUpdatePasswordUpdatesSerializedData()
     {
         $userId = $this->userObject->attribute('id');
 
-        $passwordHash = $this->getSerializedPasswordHash($this->userObject);
+        $user = ezuser::fetch( $userId );
+        $passwordHash = $user->PasswordHash;
```
Comment (glye) [inline tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php]: Could you explain this change, please?

**Comment by andrerom** [comment]
@pkamps can you accept / reject @glye's fixes?

File: kernel/classes/datatypes/ezuser/ezusertype.php [hunk 5]
```diff
@@ -258,10 +258,22 @@ function storeObjectAttribute( $contentObjectAttribute )
             }
 
             // saving information in the object attribute data_text field to simulate a draft
-            $contentObjectAttribute->setAttribute( 'data_text', $this->serializeDraft( $user ) );
+            // only if the object version is a draft (status == 0)
+            if(
+                $user->Login &&
+                $contentObjectAttribute->attribute( 'object_version' )->attribute( 'status' ) == 0
```
[the remark this answers is not in the document]
Comment (pkamps) [reply kernel/classes/datatypes/ezuser/ezusertype.php]: Hope you don't mind if I ignore it. Maybe it's a string and then it would hurt to be extra strict here.

File: kernel/user/ezuseroperationcollection.php [hunk 6]
```diff
@@ -297,7 +297,7 @@ static public function activation( $userID, $userHash, $enableUser = false )
     }
 
    /**
-     * Change user password
+     * Change user password for an existing user
```
[the remark this answers is not in the document]
Comment (pkamps) [reply kernel/user/ezuseroperationcollection.php]: I reverted my changes.

File: tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php (hunk 3 above)
Comment (pkamps) [reply tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php]: Those are required.

File: tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php (hunk 4 above)
Comment (pkamps) [reply tests/tests/kernel/datatypes/ezuser/ezusertype_regression.php]: getSerializedPasswordHash was checking the 'data_text' value of the datatype. But the function eZUserOperationCollection::password is just updating the ezuser DB table. I changed the the test to more directly check the eZUserOperationCollection::password function.

The 'data_text' value in the datatype is only a temporarily place to store user data for draft versions. If an admin creates a new users, you don't want to have the draft data in the ezuser DB table right away.

File: kernel/classes/datatypes/ezuser/ezusertype.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (glye) [reply kernel/classes/datatypes/ezuser/ezusertype.php]: We ought to be strict, strictly speaking ;) eZContentObjectVersion says it's an integer. But I see many cases where we aren't strict, so I guess this can pass if others agree.

Comment (glye) [inline kernel/classes/datatypes/ezuser/ezusertype.php]: On the other hand, it makes sense here to use `eZContentObjectVersion::STATUS_DRAFT` instead of just `0`. Same in the comment above it.

Comment (glye) [reply kernel/classes/datatypes/ezuser/ezusertype.php]: The comment could even be removed if we use the constant, since that makes the intention clear.

Comment (pkamps) [reply kernel/classes/datatypes/ezuser/ezusertype.php]: That's a good point. I changed it now.
